client/allocrunner/taskrunner/device_hook.go

```diff
@@ -0,0 +1,95 @@
+package taskrunner
+
+import (
+	"context"
+	"fmt"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	"github.com/hashicorp/nomad/client/devicemanager"
+	"github.com/hashicorp/nomad/plugins/device"
+	"github.com/hashicorp/nomad/plugins/drivers"
+)
+
+// deviceHook is used to retrieve device mounting information.
+type deviceHook struct {
+	logger log.Logger
+	dm     devicemanager.Manager
+}
+
+func newDeviceHook(dm devicemanager.Manager, logger log.Logger) *deviceHook {
+	h := &deviceHook{
+		dm: dm,
+	}
+	h.logger = logger.Named(h.Name())
+	return h
+}
+
+func (*deviceHook) Name() string {
+	return "devices"
+}
+
+func (h *deviceHook) Prestart(ctx context.Context, req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
+	if len(req.TaskResources.Devices) == 0 {
+		resp.Done = true
+		return nil
+	}
+
+	// Capture the responses
+	var reservations []*device.ContainerReservation
+	for _, req := range req.TaskResources.Devices {
+		// Ask the device manager for the reservation information
+		res, err := h.dm.Reserve(req)
+		if err != nil {
+			return fmt.Errorf("failed to reserve device %s: %v", req.ID(), err)
```

Comment:
Should we aim for some atomicity here?  In case of mid-iteration failure, should we unreserve earlier elements in loop?

Reply:
Not sure. Created an issue to address this post merge.